File: C++/BOJ_2178_미로 탐색.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

typedef vector <vector <int>> vvi;

const int dx[4] = { 1, -1, 0, 0 };
const int dy[4] = { 0, 0, 1, -1 };

int bfs(int start_x, int start_y, int column, int row, vvi table)
{
    int x, y, nx, ny;
    queue <pair <int, int>> que;
    vvi check(column, vector <int> (row)); // use this vector as displaying blocks and check

    que.push({ start_x, start_y });
    check[start_x][start_y] = 1;

    while (!que.empty()) {
        x = que.front().first; y = que.front().second;
        que.pop();

        if (x == column - 1 && y == row - 1) { // if now pos is destination
            return check[column - 1][row - 1];
        }

        for (int i = 0; i < 4; i++) {
            nx = x + dx[i]; ny = y + dy[i];
            if (nx >= 0 && nx < column && ny >= 0 && ny < row) {
                if (!check[nx][ny] && table[nx][ny]) {
                    check[nx][ny] = check[x][y] + 1;
                    que.push({ nx, ny });
                }
            }
        }
    }
    return -1;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int column, row;
    string temp;
    vvi table;

    cin >> column >> row;
    table.assign(column, vector <int>(row));

    for (int i = 0; i < column; i++) {
        cin >> temp;
        for (int j = 0; j < row; j++) {
            table[i][j] = (int)temp[j] - 48;
        }
    }

    cout << bfs(0, 0, column, row, table) << "\n";

    return 0;
}

File: Python/BOJ_2178_미로 탐색.py
import sys

def bfs(x, y, count):
    queue = [[x, y, count]]

    while queue:
        x, y, count = queue[0][0], queue[0][1], queue[0][2]
        del queue[0]

        for i in range(4):
            new_x = x + dx[i]; new_y = y + dy[i]

            if (new_x, new_y) == (row - 1, column - 1):
                print(count + 1)
                queue = []
                break
            if 0 <= new_x < row and 0 <= new_y < column:
                if table[new_x][new_y] == '1' and (new_x, new_y) not in visited:
                    queue.append([new_x, new_y, count + 1])
                    visited.append((new_x, new_y))

row, column = map(int, sys.stdin.readline().split())
table = []
for i in range(row):
    table.append(sys.stdin.readline().rstrip())

dx = [0, 0, 1, -1]; dy = [1, -1, 0, 0]
visited = [(0, 0)]
bfs(0, 0, 1)